cpp/src/arrow/compute/exec/benchmark_scripts/table_generation/datagen.py

```diff
@@ -0,0 +1,868 @@
+        """
+
+        begin_date = self._normalize_date(begin_date, date.min)
+        end_date = self._normalize_date(end_date, date.max)
+        if calendar is None:
+            calendar = SimpleMidnightCalendar()
+        offset_ns = to_offset(freq).nanos
+        if time_dist not in ["constant", "uniform"]:
+            raise ValueError("invalid time_dist: " + str(time_dist))
+        if dists is None:
+            dists = []
+        batch_size = max(1, batch_size)
+
+        schema = pa.schema(
+                [("time", pa.timestamp('ns', 'UTC'))] +
+                [(name, _arrow_type(dist.dtype)) for name, dist in dists]
+        )
+
+        uniform_time_dist = (time_dist == "uniform")
+        rng = np.random.default_rng(seed=seed)
+        one_day = timedelta(days=1)
+        curr_date = begin_date
+        curr_date_ns = self._date_to_midnight_ns(curr_date)
+        while True:
+            if calendar.is_business_day(curr_date):
+                begin_time_ns = (
+                        curr_date_ns + calendar.business_begin_ns(curr_date))
+                end_time_ns = (
+                        curr_date_ns + calendar.trading_end_ns(curr_date))
+                curr_offset_ns = min(offset_ns, end_time_ns - begin_time_ns + 1)
+                while begin_time_ns <= end_time_ns:
+                    times_num = 1 + (
+                            (end_time_ns - begin_time_ns) // curr_offset_ns)
+                    batch_num = min(times_num, batch_size)
+                    curr_array_size = (num_ids * batch_num,)
+                    batch_end_time_ns = (
+                            begin_time_ns + (batch_num - 1) * curr_offset_ns)
+
+                    time_ns_value_array = np.arange(
+                            begin_time_ns, batch_end_time_ns + 1,
+                            curr_offset_ns, dtype=np.uint64)
+                    time_ns_array = np.repeat(
+                        time_ns_value_array, num_ids).astype('datetime64[ns]')
+                    if uniform_time_dist:
+                        dist_offset_ns = min(curr_offset_ns,
+                                end_time_ns - begin_time_ns)
+                        time_ns_array += rng.integers(low=0,
+                                high=dist_offset_ns, size=curr_array_size,
+                                dtype=np.uint64)
+                        time_ns_array = np.sort(time_ns_array)
+                    arrays = ([time_ns_array] +
+                        [dist(rng, num_ids, batch_num) for _, dist in dists])
+                    parrays = [pa.array(a) for a in arrays]
+
+                    batch = pa.RecordBatch.from_arrays(parrays, schema=schema)
+                    yield batch
+
+                    begin_time_ns += batch_num * curr_offset_ns
+            if curr_date == end_date:
+                break
+            curr_date += one_day
+            curr_date_ns = self._date_to_midnight_ns(curr_date)
+
+class TableMDGenerator:
```

Comment:
Why are there so many TableXyzGenerator types?  What is common / different between them all?